File: src/main/kotlin/org/ninelym/cognitive/embodiment/ROSBinding.kt
package org.ninelym.cognitive.embodiment

import org.ninelym.cognitive.CognitiveTensor
import org.ninelym.cognitive.api.AgentType
import org.ninelym.cognitive.api.CognitiveMeshAPI

/**
 * ROS (Robot Operating System) Embodiment Binding
 * 
 * Provides integration layer between cognitive mesh and ROS-based robotic systems.
 * Supports topics, services, actions, and tf (transforms) framework.
 */
class ROSBinding(private val meshAPI: CognitiveMeshAPI) {
    
    private val rosAgents = mutableMapOf<String, ROSAgent>()
    private val activeTopics = mutableMapOf<String, ROSTopicStream>()
    
    /**
     * Connect ROS node to cognitive mesh
     */
    fun connectROSNode(
        nodeName: String,
        robotCapabilities: List<String>,
        rosVersion: String = "ROS2",
        namespace: String = "/"
    ): ROSConnection {
        
        val registrationRequest = org.ninelym.cognitive.api.AgentRegistrationRequest(
            name = "ROS-$nodeName",
            type = AgentType.ROS_ROBOT,
            capabilities = robotCapabilities + listOf(
                "navigation", "manipulation", "perception", "localization",
                "mapping", "planning", "control", "tf_transforms"
            ),
            autonomyLevel = 0.8f
        )
        
        val response = meshAPI.registerAgent(registrationRequest)
        
        val rosAgent = ROSAgent(
            agentId = response.agentId,
            nodeName = nodeName,
            namespace = namespace,
            rosVersion = rosVersion,
            topics = mutableMapOf(),
            services = mutableMapOf(),
            isConnected = response.success,
            lastHeartbeat = System.currentTimeMillis()
        )
        
        rosAgents[response.agentId] = rosAgent
        
        return ROSConnection(
            agentId = response.agentId,
            nodeName = nodeName,
            success = response.success,
            apiEndpoints = response.apiEndpoints,
            rosSpecificEndpoints = ROSEndpoints(
                topics = "/ros/${response.agentId}/topics",
                services = "/ros/${response.agentId}/services",
                actions = "/ros/${response.agentId}/actions",
                tf = "/ros/${response.agentId}/tf",
                diagnostics = "/ros/${response.agentId}/diagnostics"
            )
        )
    }
    
    /**
     * Publish data to ROS topic through cognitive mesh
     */
    fun publishToTopic(
        agentId: String,
        topicName: String,
        messageType: String,
        messageData: Map<String, Any>,
        priority: Float = 0.5f
    ): ROSPublishResponse {
        
        val agent = rosAgents[agentId] 
            ?: return ROSPublishResponse(false, "ROS agent not found")
        
        // Create cognitive tensor for ROS message
        val messageTensor = CognitiveTensor(
            modality = calculateModalityForTopic(topicName, messageType),
            depth = calculateMessageComplexity(messageData),
            context = calculateTopicContext(topicName),
            salience = priority,
            autonomyIndex = 0.8f
        )
        
        // Submit as sensor data
        val sensorRequest = org.ninelym.cognitive.api.SensorDataRequest(
            sensorType = "ros-topic-$topicName",
            data = ROSMessage(topicName, messageType, messageData),
            modality = messageTensor.modality,
            processingDepth = messageTensor.depth,
            contextRelevance = messageTensor.context,
            importance = messageTensor.salience
        )
        
        val response = meshAPI.submitSensorData(agentId, sensorRequest)
        
        if (response.success) {
            // Update topic tracking
            val topicId = "${agentId}-${topicName}"
            activeTopics[topicId] = ROSTopicStream(
                agentId = agentId,
                topicName = topicName,
                messageType = messageType,
                lastPublish = System.currentTimeMillis(),
                messageCount = (activeTopics[topicId]?.messageCount ?: 0) + 1,
                isActive = true
            )
            
            agent.lastHeartbeat = System.currentTimeMillis()
        }
        
        return ROSPublishResponse(
            success = response.success,
            message = response.message,
            cognitiveProcessing = response.cognitiveResponse,
            topicId = topicName,
            messageId = System.currentTimeMillis().toString()
        )
    }
    
    /**
     * Subscribe to cognitive recommendations for ROS actions
     */
    fun getROSActionRecommendations(agentId: String): List<ROSActionRecommendation> {
        val actionsResponse = meshAPI.getAvailableActions(agentId)
        
        return actionsResponse.actions.mapNotNull { action ->
            when (action.id) {
                "ros-move" -> ROSActionRecommendation(
                    actionType = ROSActionType.NAVIGATION,
                    topicName = "/cmd_vel",
                    messageType = "geometry_msgs/Twist",
                    parameters = action.parameters,
                    priority = action.priority,
                    description = action.description
                )
                "ros-publish" -> ROSActionRecommendation(
                    actionType = ROSActionType.COMMUNICATION,
                    topicName = action.parameters["topic"] ?: "/status",
                    messageType = "std_msgs/String",
                    parameters = action.parameters,
                    priority = action.priority,
                    description = action.description
                )
                else -> null
            }
        }
    }
    
    /**
     * Update robot pose/odometry data
     */
    fun updateRobotPose(
        agentId: String,
        pose: ROSPose,
        frameId: String = "base_link",
        childFrameId: String = "odom"
    ): Boolean {
        
        val poseTensor = CognitiveTensor(
            modality = 0.9f, // Spatial awareness is high modality
            depth = 0.7f,
            context = 0.8f, // Pose is contextually important
            salience = calculatePoseSalience(pose),
            autonomyIndex = 0.8f
        )
        
        val sensorRequest = org.ninelym.cognitive.api.SensorDataRequest(
            sensorType = "ros-pose",
            data = ROSOdometry(pose, frameId, childFrameId),
            modality = poseTensor.modality,
            processingDepth = poseTensor.depth,
            contextRelevance = poseTensor.context,
            importance = poseTensor.salience
        )
        
        return meshAPI.submitSensorData(agentId, sensorRequest).success
    }
    
    /**
     * Process sensor data (laser scan, camera, etc.)
     */
    fun processSensorData(
        agentId: String,
        sensorType: ROSSensorType,
        sensorData: Map<String, Any>
    ): ROSSensorResponse {
        
        val sensorTensor = CognitiveTensor(
            modality = when (sensorType) {
                ROSSensorType.LASER_SCAN -> 0.8f
                ROSSensorType.CAMERA_IMAGE -> 0.9f
                ROSSensorType.POINT_CLOUD -> 0.9f
                ROSSensorType.IMU -> 0.7f
                ROSSensorType.GPS -> 0.6f
                ROSSensorType.SONAR -> 0.7f
            },
            depth = when (sensorType) {
                ROSSensorType.POINT_CLOUD -> 0.9f
                ROSSensorType.CAMERA_IMAGE -> 0.8f
                else -> 0.6f
            },
            context = 0.8f,
            salience = calculateSensorSalience(sensorType, sensorData),
            autonomyIndex = 0.8f
        )
        
        val sensorRequest = org.ninelym.cognitive.api.SensorDataRequest(
            sensorType = "ros-sensor-${sensorType.name.lowercase()}",
            data = sensorData,
            modality = sensorTensor.modality,
            processingDepth = sensorTensor.depth,
            contextRelevance = sensorTensor.context,
            importance = sensorTensor.salience
        )
        
        val response = meshAPI.submitSensorData(agentId, sensorRequest)
        
        return ROSSensorResponse(
            processed = response.success,
            sensorType = sensorType,
            cognitiveAnalysis = response.cognitiveResponse,
            navigationRecommendations = if (response.success) {
                generateNavigationRecommendations(sensorType, sensorData)
            } else emptyList(),
            perceptionInsights = if (response.success) {
                generatePerceptionInsights(sensorType, sensorData)
            } else emptyList()
        )
    }
    
    /**
     * Execute ROS action based on cognitive recommendation
     */
    fun executeROSAction(
        agentId: String,
        action: ROSActionRecommendation
    ): ROSActionResult {
        
        // This would interface with actual ROS action servers
        // For now, we simulate the action execution
        
        val startTime = System.currentTimeMillis()
        val success = simulateROSActionExecution(action)
        val executionTime = System.currentTimeMillis() - startTime
        
        return ROSActionResult(
            actionType = action.actionType,
            success = success,
            executionTime = executionTime,
            result = if (success) "Action completed successfully" else "Action failed",
            feedback = "Cognitive mesh guided action execution"
        )
    }
    
    private fun calculateModalityForTopic(topicName: String, messageType: String): Float {
        return when {
            topicName.contains("camera") || messageType.contains("Image") -> 0.9f
            topicName.contains("scan") || messageType.contains("LaserScan") -> 0.8f
            topicName.contains("cmd_vel") || messageType.contains("Twist") -> 0.7f
            topicName.contains("odom") -> 0.8f
            else -> 0.6f
        }
    }
    
    private fun calculateMessageComplexity(messageData: Map<String, Any>): Float {
        val fieldCount = messageData.size
        val hasNestedStructures = messageData.values.any { it is Map<*, *> || it is List<*> }
        
        var complexity = fieldCount.toFloat() / 20f
        if (hasNestedStructures) complexity += 0.3f
        
        return complexity.coerceIn(0.1f, 1.0f)
    }
    
    private fun calculateTopicContext(topicName: String): Float {
        return when {
            topicName.contains("emergency") || topicName.contains("alert") -> 0.9f
            topicName.contains("navigation") || topicName.contains("cmd") -> 0.8f
            topicName.contains("status") || topicName.contains("diagnostic") -> 0.7f
            else -> 0.6f
        }
    }
    
    private fun calculatePoseSalience(pose: ROSPose): Float {
        // Calculate salience based on pose change magnitude
        val positionMagnitude = Math.sqrt(
            (pose.position.x * pose.position.x +
             pose.position.y * pose.position.y +
             pose.position.z * pose.position.z).toDouble()
        ).toFloat()
        
        return (positionMagnitude / 10f).coerceIn(0.1f, 1.0f)
    }
    
    private fun calculateSensorSalience(
        sensorType: ROSSensorType,
        sensorData: Map<String, Any>
    ): Float {
        return when (sensorType) {
            ROSSensorType.LASER_SCAN -> {
                // Higher salience for closer obstacles
                val minRange = (sensorData["ranges"] as? List<*>)
                    ?.filterIsInstance<Float>()
                    ?.minOrNull() ?: 10f
                (10f - minRange.coerceIn(0f, 10f)) / 10f
            }
            ROSSensorType.CAMERA_IMAGE -> 0.8f // Visual data is generally high salience
            ROSSensorType.IMU -> {
                // Higher salience for significant acceleration/rotation
                0.6f // Simplified
            }
            else -> 0.5f
        }
    }
    
    private fun generateNavigationRecommendations(
        sensorType: ROSSensorType,
        sensorData: Map<String, Any>
    ): List<String> {
        return when (sensorType) {
            ROSSensorType.LASER_SCAN -> listOf(
                "ObstacleAvoidance",
                "PathReplanning",
                "SpeedReduction"
            )
            ROSSensorType.CAMERA_IMAGE -> listOf(
                "VisualNavigation",
                "LandmarkDetection",
                "SemanticMapping"
            )
            ROSSensorType.GPS -> listOf(
                "GlobalLocalization", 
                "RouteOptimization"
            )
            else -> emptyList()
        }
    }
    
    private fun generatePerceptionInsights(
        sensorType: ROSSensorType,
        sensorData: Map<String, Any>
    ): List<String> {
        return when (sensorType) {
            ROSSensorType.CAMERA_IMAGE -> listOf(
                "ObjectDetection",
                "SceneUnderstanding",
                "DepthEstimation"
            )
            ROSSensorType.POINT_CLOUD -> listOf(
                "3DMapping",
                "SurfaceAnalysis", 
                "ObjectSegmentation"
            )
            ROSSensorType.LASER_SCAN -> listOf(
                "2DMapping",
                "WallDetection",
                "CornerExtraction"
            )
            else -> emptyList()
        }
    }
    
    private fun simulateROSActionExecution(action: ROSActionRecommendation): Boolean {
        // Simulate action execution with some randomness for demo purposes
        return Math.random() > 0.1 // 90% success rate
    }
}

/**
 * ROS-specific data structures
 */

data class ROSConnection(
    val agentId: String,
    val nodeName: String,
    val success: Boolean,
    val apiEndpoints: org.ninelym.cognitive.api.ApiEndpoints,
    val rosSpecificEndpoints: ROSEndpoints
)

data class ROSEndpoints(
    val topics: String,
    val services: String,
    val actions: String,
    val tf: String,
    val diagnostics: String
)

data class ROSAgent(
    val agentId: String,
    val nodeName: String,
    val namespace: String,
    val rosVersion: String,
    val topics: MutableMap<String, String>, // topic name -> message type
    val services: MutableMap<String, String>,
    var isConnected: Boolean,
    var lastHeartbeat: Long
)

data class ROSTopicStream(
    val agentId: String,
    val topicName: String,
    val messageType: String,
    var lastPublish: Long,
    var messageCount: Long,
    var isActive: Boolean
)

data class ROSMessage(
    val topicName: String,
    val messageType: String,
    val data: Map<String, Any>
)

data class ROSPose(
    val position: ROSPoint,
    val orientation: ROSQuaternion
)

data class ROSPoint(
    val x: Float,
    val y: Float,
    val z: Float
)

data class ROSQuaternion(
    val x: Float,
    val y: Float,
    val z: Float,
    val w: Float
)

data class ROSOdometry(
    val pose: ROSPose,
    val frameId: String,
    val childFrameId: String
)

enum class ROSSensorType {
    LASER_SCAN,
    CAMERA_IMAGE,
    POINT_CLOUD,
    IMU,
    GPS,
    SONAR
}

enum class ROSActionType {
    NAVIGATION,
    MANIPULATION,
    PERCEPTION,
    COMMUNICATION,
    CONTROL
}

data class ROSActionRecommendation(
    val actionType: ROSActionType,
    val topicName: String,
    val messageType: String,
    val parameters: Map<String, String>,
    val priority: Float,
    val description: String
)

data class ROSPublishResponse(
    val success: Boolean,
    val message: String,
    val cognitiveProcessing: org.ninelym.cognitive.api.CognitiveProcessingResponse? = null,
    val topicId: String = "",
    val messageId: String = ""
)

data class ROSSensorResponse(
    val processed: Boolean,
    val sensorType: ROSSensorType,
    val cognitiveAnalysis: org.ninelym.cognitive.api.CognitiveProcessingResponse?,
    val navigationRecommendations: List<String>,
    val perceptionInsights: List<String>
)

data class ROSActionResult(
    val actionType: ROSActionType,
    val success: Boolean,
    val executionTime: Long,
    val result: String,
    val feedback: String
)

File: python/helpers/phase4/embodiment/ros_binding.py
"""
ROS Binding for Cognitive Mesh

Integrates with Robot Operating System (ROS/ROS2):
- Topic publishing/subscribing
- Service calls and actions
- Transform (tf) integration
- Sensor data processing
- Navigation recommendations
"""

from dataclasses import asdict, dataclass
from enum import Enum
from typing import Any, Dict, List, Optional
import time


# ============================================================================
# Enums and Data Classes
# ============================================================================


class ROSSensorType(Enum):
    """ROS sensor types."""
    
    LASER_SCAN = "LASER_SCAN"
    CAMERA_IMAGE = "CAMERA_IMAGE"
    POINT_CLOUD = "POINT_CLOUD"
    IMU = "IMU"
    GPS = "GPS"
    SONAR = "SONAR"


@dataclass
class ROSConnection:
    """ROS node connection information."""
    
    agent_id: str
    success: bool
    message: str
    node_name: str
    ros_version: str
    capabilities: List[str]
    timestamp: float


# ============================================================================
# ROS Binding
# ============================================================================


class ROSBinding:
    """
    ROS/ROS2 integration binding for cognitive mesh.
    
    Provides topic publishing, sensor processing, and
    cognitive navigation recommendations for robots.
    """
    
    def __init__(self, mesh_api):
        """
        Initialize ROS binding.
        
        Args:
            mesh_api: CognitiveMeshAPI instance
        """
        self.mesh_api = mesh_api
        self.topics: Dict[str, Dict[str, Any]] = {}
        self.services: Dict[str, List[str]] = {}
        self.sensor_data: Dict[str, List[Dict[str, Any]]] = {}
        
        print("[ROSBinding] Initialized")
    
    def connect_ros_node(
        self,
        node_name: str,
        robot_capabilities: List[str],
        ros_version: str = "ROS2",
    ) -> ROSConnection:
        """
        Connect a ROS node to the mesh.
        
        Args:
            node_name: Name of the ROS node
            robot_capabilities: List of robot capabilities
            ros_version: ROS version (ROS or ROS2)
        
        Returns:
            ROSConnection with agent info
        """
        # Register with mesh API
        registration = self.mesh_api.register_agent(
            client_name=node_name,
            capabilities=robot_capabilities,
            agent_type="ros",
        )
        
        if registration.success:
            # Initialize topic and service tracking
            self.topics[registration.agent_id] = {}
            self.services[registration.agent_id] = []
            self.sensor_data[registration.agent_id] = []
            
            print(f"[ROSBinding] Connected node: {node_name} ({registration.agent_id})")
        
        return ROSConnection(
            agent_id=registration.agent_id,
            success=registration.success,
            message=registration.message,
            node_name=node_name,
            ros_version=ros_version,
            capabilities=robot_capabilities,
            timestamp=time.time(),
        )
    
    def publish_to_topic(
        self,
        agent_id: str,
        topic_name: str,
        message_type: str,
        message_data: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Publish data to a ROS topic.
        
        Args:
            agent_id: Agent identifier
            topic_name: ROS topic name (e.g., /cmd_vel)
            message_type: ROS message type (e.g., geometry_msgs/Twist)
            message_data: Message data dictionary
        
        Returns:
            Response with publishing status
        """
        if agent_id not in self.topics:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Store topic data
        self.topics[agent_id][topic_name] = {
            "message_type": message_type,
            "message_data": message_data,
            "timestamp": time.time(),
        }
        
        # Submit as sensor data to mesh
        sensor_data = {
            "topicName": topic_name,
            "messageType": message_type,
            "messageData": message_data,
        }
        
        self.mesh_api.submit_sensor_data(
            agent_id=agent_id,
            sensor_type="ros_topic",
            sensor_data=sensor_data,
        )
        
        return {
            "success": True,
            "agent_id": agent_id,
            "topic_name": topic_name,
            "message_type": message_type,
            "published": True,
            "cognitive_feedback": self._get_topic_feedback(topic_name, message_data),
        }
    
    def process_sensor_data(
        self,
        agent_id: str,
        sensor_type: ROSSensorType,
        sensor_data: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Process ROS sensor data and generate cognitive insights.
        
        Args:
            agent_id: Agent identifier
            sensor_type: Type of ROS sensor
            sensor_data: Sensor data dictionary
        
        Returns:
            Response with cognitive analysis
        """
        if agent_id not in self.topics:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Store sensor data
        sensor_entry = {
            "sensor_type": sensor_type.value,
            "sensor_data": sensor_data,
            "timestamp": time.time(),
        }
        self.sensor_data[agent_id].append(sensor_entry)
        
        # Keep only last 20 sensor readings
        if len(self.sensor_data[agent_id]) > 20:
            self.sensor_data[agent_id] = self.sensor_data[agent_id][-20:]
        
        # Submit to mesh API
        self.mesh_api.submit_sensor_data(
            agent_id=agent_id,
            sensor_type=f"ros_sensor_{sensor_type.value.lower()}",
            sensor_data=sensor_data,
        )
        
        # Generate cognitive insights
        insights = self._analyze_sensor_data(sensor_type, sensor_data)
        
        return {
            "success": True,
            "agent_id": agent_id,
            "sensor_type": sensor_type.value,
            "cognitive_insights": insights,
            "recommendations": self._get_sensor_recommendations(sensor_type, insights),
        }
    
    def get_navigation_recommendations(
        self,
        agent_id: str,
        current_pose: Dict[str, Any],
        goal_pose: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Get cognitive navigation recommendations.
        
        Args:
            agent_id: Agent identifier
            current_pose: Current robot pose
            goal_pose: Goal pose
        
        Returns:
            Navigation recommendations
        """
        if agent_id not in self.topics:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Calculate simple navigation metrics
        dx = goal_pose.get("x", 0) - current_pose.get("x", 0)
        dy = goal_pose.get("y", 0) - current_pose.get("y", 0)
        distance = (dx**2 + dy**2)**0.5
        
        import math
        target_angle = math.atan2(dy, dx)
        
        recommendations = []
        
        if distance > 5.0:
            recommendations.append({
                "action": "plan_global_path",
                "priority": "high",
                "params": {"algorithm": "A*", "resolution": 0.05},
            })
        elif distance > 1.0:
            recommendations.append({
                "action": "adjust_local_path",
                "priority": "medium",
                "params": {"lookahead": 0.5},
            })
        else:
            recommendations.append({
                "action": "final_approach",
                "priority": "high",
                "params": {"precision": 0.1},
            })
        
        return {
            "success": True,
            "agent_id": agent_id,
            "distance_to_goal": distance,
            "target_angle": target_angle,
            "recommendations": recommendations,
            "cognitive_state": "navigating" if distance > 0.5 else "arrived",
        }
    
    def call_ros_service(
        self,
        agent_id: str,
        service_name: str,
        service_type: str,
        service_request: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Call a ROS service.
        
        Args:
            agent_id: Agent identifier
            service_name: ROS service name
            service_type: ROS service type
            service_request: Service request data
        
        Returns:
            Service response
        """
        if agent_id not in self.topics:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Track service call
        if service_name not in self.services[agent_id]:
            self.services[agent_id].append(service_name)
        
        # Generate mock response (in real implementation, would call actual ROS service)
        return {
            "success": True,
            "agent_id": agent_id,
            "service_name": service_name,
            "service_type": service_type,
            "service_response": {
                "result": "success",
                "data": service_request,
            },
            "cognitive_assessment": "Service call processed successfully",
        }
    
    def _get_topic_feedback(
        self,
        topic_name: str,
        message_data: Dict[str, Any],
    ) -> str:
        """Get cognitive feedback for topic publishing."""
        if "/cmd_vel" in topic_name:
            linear = message_data.get("linear", {})
            angular = message_data.get("angular", {})
            
            if abs(linear.get("x", 0)) > 0.5:
                return "High speed command - ensure obstacle detection is active"
            elif abs(angular.get("z", 0)) > 0.5:
                return "Sharp turn command - monitor for stability"
            else:
                return "Safe velocity command"
        
        return "Topic data published successfully"
    
    def _analyze_sensor_data(
        self,
        sensor_type: ROSSensorType,
        sensor_data: Dict[str, Any],
    ) -> Dict[str, Any]:
        """Analyze sensor data for cognitive insights."""
        insights = {
            "sensor_type": sensor_type.value,
            "data_quality": "good",
            "anomalies": [],
        }
        
        if sensor_type == ROSSensorType.LASER_SCAN:
            ranges = sensor_data.get("ranges", [])
            if ranges:
                min_range = min(ranges)
                if min_range < 0.5:
                    insights["anomalies"].append("Obstacle detected nearby")
                    insights["data_quality"] = "warning"
        
        elif sensor_type == ROSSensorType.IMU:
            # Check for excessive acceleration
            pass  # Simplified for this implementation
        
        return insights
    
    def _get_sensor_recommendations(
        self,
        sensor_type: ROSSensorType,
        insights: Dict[str, Any],
    ) -> List[Dict[str, Any]]:
        """Get recommendations based on sensor insights."""
        recommendations = []
        
        if insights.get("data_quality") == "warning":
            if sensor_type == ROSSensorType.LASER_SCAN:
                recommendations.append({
                    "action": "reduce_velocity",
                    "priority": "high",
                    "reason": "Obstacle detected",
                })
        
        return recommendations
